### PR TITLE
Allow AuthToken to accept SHA256 or SHA1 hashes

### PR DESCRIPTION
**Pull request by KludgeKML**
⚠️  This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

- Makes SHA256 tokens the default again (as per Rails 7 defaults)
- Adds a rotate to the decrypter that allows it to accept SHA1 tokens as well as SHA256 tokens.

This will allow it to accept the current SHA1 tokens from email-alert-api, and the SHA256 tokens from when email-alert-api is updated properly.

https://trello.com/c/Iz4TyAzx/1321-fixes-for-email-alert-api-email-alert-frontend-update-incident

Note - due to the way email-alert-api is set up, not currently possible to test this end to end. Has been tested by generating a token in the email-alert-api integration console:

`token = AuthTokenGeneratorService.call(hello: 'world')`

...then checking it decrypts on the email-alert-frontend integration console:

```
irb(main):007:0> token = AuthToken.new(data)
=> #<AuthToken:0x00007f23de69fb78 @token="1xZjrFbz4korhQywWP9H2nycQBNdySnzq...
irb(main):008:0> token.valid?
=> true
irb(main):009:0> token.data
=> {:hello=>"world"}
```

### CODE REVIEW

File: spec/models/auth_token_spec.rb
Comment (ChrisBAshton) [inline spec/models/auth_token_spec.rb:12]: These two `returns nil` tests don't give me much confidence as they don't pass the `data: { a: "b" }` part of the hash, so even without an `expiry: 0`, I'd expect `token.data` to be `nil` 🤔 
Could you add the `data` property to explicitly test this?

**Review by ChrisBAshton — COMMENTED** [review]
Well done for coming up with a solution to this thorny problem!
Just so I'm clear, we'd expect to be able to revert this PR within say an hour of updating email-alert-api, right?

Nitpick: could you copy all the useful PR description into the commit message itself? Saves a future developer having to navigate to the PR in Git if their local IDE shows git commit information.

**Comment by KludgeKML** [comment]
Reverting - the links are advertised as being valid for 7 days, so I think we'd want to wait a week between updating email-alert-api and updating this (it wouldn't strictly be a revert because currently email-alert-frontend is only accepting SHA1 tokens, and we'd want it to move to only accepting SHA256 once we were sure all the SHA1 tokens from email-alert-api were expired).

Details in commit - yeah, I can update the commit with the notes.

File: spec/models/auth_token_spec.rb
Comment (KludgeKML) [reply spec/models/auth_token_spec.rb:12]: The default data is actually an empty hash, so if you set the expiry to (say) 1000, or remove it entirely, the test will fail - it's only nil if the token is invalid for some reason. BUT! there's actually a nicer way of doing this - I can expand the test suite and move these tests into a test for the #valid? method, which would be a bit more explicit.

Comment (KludgeKML) [reply spec/models/auth_token_spec.rb]: Have updated the tests, see if the current change seems less worrying.

File: spec/models/auth_token_spec.rb
Comment (ChrisBAshton) [inline spec/models/auth_token_spec.rb:6]: Nitpick: we tend to avoid using `context` blocks unless there's more than one `it` block inside it. In this case, we'd usually just do:

```suggestion
    it "returns the data hash when given a valid SHA1 token" do
```

**Review by ChrisBAshton — APPROVED** [review]
LGTM 👍 probably worth building the branch to the relevant env to check it works, before merging.

One nitpick comment but doesn't block the PR.

**Review by ChrisBAshton — APPROVED** [review]

**Pull request merged**
